### PR TITLE
feat: provide ability to relocate ipc.socket file

### LINKED ISSUE #1269: Nucleus: Provide ability to relocate ipc.socket file
**Feature Description**
Ability to relocate the `$GG_ROOT/ipc.socket` to another location on the filesystem, via initial configuration file.

**Use Case**
When running Greengrass in a container, and also running components as components that need to utilize the IPC, having `ipc.socket` in the `$GG_ROOT` location requires mapping the _entire_ directory structure (and sub-directories) as a docker volume to then make it accessible to the sibling component containers.

By allowing the file to be located in another directory, docker `volume` commands could then be specific to only mount that specific location.

**Proposed Solution**
Change the Nucleus to check for an initial config setting for the `ipc.socket` file location (e.g., `/tmp/greengrass/ipc.socket` or `/var/run/greengrass/ipc.socket`). During configuration, Greengrass will create the file within that directory, and then the container running the Greengrass Nucleus can make that available via a volume.

Components that run as containers on the same host can then map the exported volume to their namespace, and then use `AWS_GG_NUCLEUS_DOMAIN_SOCKET_FILEPATH_FOR_COMPONENT` to refer to the location within their containers for IPC use (along with the `SVCUID`). 

**Other**
Feature would reduce blast radius of a component by removing access to the full `GG_ROOT` directory structure.

* [ ] :wave: I may be able to implement this feature request
* [X] :warning: This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by Mirrorxxi**
**Issue #, if available:**
Resolves #1269.

**Description of changes:**
Adds a new configuration under the system namespace to define where the ipc socket file should be created on Linux.

```yaml
system:
  ipcSocketPath: /var/run/gg.socket
```

**Why is this change necessary:**

**How was this change tested:**
- [x] Updated or added new unit tests.
- [x] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [ ] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/easysetup/DeviceProvisioningHelper.java [hunk 1]
```diff
@@ -360,6 +360,7 @@ public void updateKernelConfigWithIotConfiguration(Kernel kernel, ThingInfo thin
         Path caFilePath = rootDir.resolve("rootCA.pem");
         Path privKeyFilePath = rootDir.resolve("privKey.key");
         Path certFilePath = rootDir.resolve("thingCert.crt");
+        Path ipcSocketPath = rootDir.resolve("ipc.socket");
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/easysetup/DeviceProvisioningHelper.java:363]: This code should not be here. Revert all changes in this file.

File: src/main/java/com/aws/greengrass/ipc/IPCEventStreamService.java [hunk 2]
```diff
@@ -60,12 +61,17 @@ public class IPCEventStreamService implements Startable, Closeable {
     private SocketOptions socketOptions;
     private EventLoopGroup eventLoopGroup;
 
+    private static DeviceConfiguration deviceConfiguration;
+
     @Inject
     IPCEventStreamService(Kernel kernel,
                           GreengrassCoreIPCService greengrassCoreIPCService,
                           Configuration config,
                           AuthenticationHandler authenticationHandler) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/ipc/IPCEventStreamService.java:70]: Add `DeviceConfiguration` to the constructor. It will then be automatically injected. Do not manually lookup the DeviceConfiguration from the Context.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java [hunk 3]
```diff
@@ -136,6 +143,16 @@ void beforeEach(ExtensionContext context) {
         ignoreExceptionOfType(context, InterruptedException.class);
     }
 
+    @Test
+    void Given_assign_path_for_ipcSocket_When_startUp_Then_ipcSocket_store_in_assigned_path() {
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:147]: this test isn't testing using a customized path. 

You must add a test case which covers the case when the user chooses to put the socket file somewhere else.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCTestUtils.java [hunk 4]
```diff
@@ -72,6 +74,11 @@ public static Kernel prepareKernelFromConfigFile(String configFile, Class testCl
         GlobalStateChangeListener listener = getListenerForServiceRunning(awaitIpcServiceLatch, serviceNames);
         kernel.getContext().addGlobalStateChangeListener(listener);
 
+        DeviceConfiguration deviceConfiguration = new DeviceConfiguration(kernel, "ThingName", "xxxxxx-ats.iot.us-east-1.amazonaws.com",
+                "xxxxxx.credentials.iot.us-east-1.amazonaws.com", "privKeyFilePath", "certFilePath", "caFilePath",
+                 "us-east-1", "roleAliasName");
+        kernel.getContext().put(DeviceConfiguration.class,deviceConfiguration);
+
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCTestUtils.java:81]: These changes should not be necessary, please remove them.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java [hunk 5]
```diff
@@ -102,7 +109,7 @@ class IPCServicesTest extends BaseITCase {
     private static GreengrassCoreIPCClient greengrassCoreIPCClient;
 
     @BeforeAll
-    static void beforeAll() throws InterruptedException, ExecutionException, IOException {
+    static void beforeAll() throws InterruptedException, ExecutionException, IOException, DeviceConfigurationException {
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:112]: remove the `throws DeviceConfigurationException`.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCPubSubTest.java [hunk 6]
```diff
@@ -9,6 +9,7 @@
 import com.aws.greengrass.authorization.Permission;
 import com.aws.greengrass.config.Topics;
 import com.aws.greengrass.config.UpdateBehaviorTree;
+import com.aws.greengrass.deployment.exceptions.DeviceConfigurationException;
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCPubSubTest.java:12]: revert the changes in this file

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCPubSubRemovalTest.java [hunk 7]
```diff
@@ -10,6 +10,7 @@
 import com.aws.greengrass.componentmanager.exceptions.PackageDownloadException;
 import com.aws.greengrass.config.Topic;
 import com.aws.greengrass.config.Topics;
+import com.aws.greengrass.deployment.exceptions.DeviceConfigurationException;
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCPubSubRemovalTest.java:13]: revert the changes in this file

File: src/main/java/com/aws/greengrass/ipc/IPCEventStreamService.java [hunk 8]
```diff
@@ -75,6 +81,7 @@ public class IPCEventStreamService implements Startable, Closeable {
     @Override
     public void startup() {
         Path rootPath = kernel.getNucleusPaths().rootPath();
+        String ipcPath = Coerce.toString(deviceConfiguration.getIpcSocketPath());
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/ipc/IPCEventStreamService.java:84]: Convert the path into a `Path`

```suggestion
        String ipcPathStr = Coerce.toString(deviceConfiguration.getIpcSocketPath());
        Path ipcPath = Utils.isEmpty(ipcPathStr) ? null : Paths.get(ipcPathStr);
```

File: src/main/java/com/aws/greengrass/ipc/IPCEventStreamService.java [hunk 9]
```diff
@@ -164,6 +171,7 @@ public void close() {
             socketOptions.close();
         }
 
-        Platform.getInstance().cleanupIpcFiles(kernel.getNucleusPaths().rootPath());
+        String ipcPath = Coerce.toString(deviceConfiguration.getIpcSocketPath());
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/ipc/IPCEventStreamService.java:174]: same thing down here, this should be a `Path`

File: src/main/java/com/aws/greengrass/util/platforms/unix/UnixPlatform.java [hunk 10]
```diff
@@ -549,17 +549,27 @@ private List<String> children(String parent, Map<String, List<String>> procMap)
         return ret;
     }
 
-    private String getIpcServerSocketAbsolutePath(Path rootPath) {
-        return rootPath.resolve(IPC_SERVER_DOMAIN_SOCKET_FILENAME).toString();
+    private String getIpcServerSocketAbsolutePath(Path rootPath, String ipcPath) {
+        if (Utils.isEmpty(ipcPath)) {
+            ipcPath = IPC_SERVER_DOMAIN_SOCKET_FILENAME;
+        }
+        return rootPath.resolve(ipcPath).toString();
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/unix/UnixPlatform.java:556]: If `ipcPath` is provided by the user, then that path is _not_ relative to the root path. The user-provided path should be used directly, without modification.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java [hunk 11]
```diff
@@ -136,6 +142,16 @@ void beforeEach(ExtensionContext context) {
...
     }
 
+    @Test
+    void Given_assign_path_for_ipcSocket_When_startUp_Then_ipcSocket_store_in_assigned_path() {
+        DeviceConfiguration deviceConfiguration = kernel.getContext().get(DeviceConfiguration.class);
+        String ipcPath = Coerce.toString(deviceConfiguration.getIpcSocketPath());
+        if (Utils.isEmpty(ipcPath)) {
+            Path rootPath = kernel.getNucleusPaths().rootPath();
+            ipcPath = rootPath.resolve("ipc.socket").toString();
+        }
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:152]: remove the `if`. This test has a well defined path

File: src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/ipc.yaml [hunk 12]
```diff
@@ -1,7 +1,20 @@
 ---
+system:
+  certificateFilePath: "/greengrass/v2/device.pem.crt"
+  privateKeyPath: "/greengrass/v2/private.pem.key"
+  rootCaPath: "/greengrass/v2/AmazonRootCA1.pem"
+  rootpath: "/greengrass/v2"
+  ipcSocketPath: "/ipcFolder/ipc.socket"
+  thingName: "GreengrassCore-yiwen-ubuntu-ec2-01"
```
Comment (MikeDombo) [inline src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/ipc.yaml:8]: remove thing name

File: src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/ipc.yaml [hunk 13]
```diff
@@ -1,7 +1,20 @@
...
 services:
   aws.greengrass.Nucleus:
+    componentType: "NUCLEUS"
+    version: "2.7.0"
     configuration:
+      awsRegion: "us-east-1"
+      iotRoleAlias: "GreengrassCoreTokenExchangeRoleAlias"
+      iotDataEndpoint: "a127t3ngyir9g6-ats.iot.us-east-1.amazonaws.com"
+      iotCredEndpoint: "c3fmfzs1ifldoe.credentials.iot.us-east-1.amazonaws.com"
```
Comment (MikeDombo) [inline src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/ipc.yaml:17]: None of these changes should be necessary. Remove them.

File: src/main/java/com/aws/greengrass/util/platforms/unix/UnixPlatform.java [hunk 14]
```diff
@@ -549,17 +549,27 @@ private List<String> children(String parent, Map<String, List<String>> procMap)
...
     }
 
     @Override
-    public String prepareIpcFilepath(Path rootPath) {
-        String ipcServerSocketAbsolutePath = getIpcServerSocketAbsolutePath(rootPath);
+    public String prepareIpcFilepath(Path rootPath, Path ipcPath) {
+        String ipcServerSocketAbsolutePath = getIpcServerSocketAbsolutePath(rootPath, ipcPath);
+
+        try {
+            Utils.createPaths(Paths.get(ipcServerSocketAbsolutePath));
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/unix/UnixPlatform.java:568]: createPaths needs a directory. This path will be the path to a file, not a directory. Get the parent path first, then create the paths.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java [hunk 15]
```diff
@@ -136,6 +142,13 @@ void beforeEach(ExtensionContext context) {
         ignoreExceptionOfType(context, InterruptedException.class);
     }
 
+    @Test
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:145]: This test does not work on Windows due to differences in how Windows works. This particular test should be skipped on Windows.

`@DisabledOnOs(OS.WINDOWS)`

**Review by MikeDombo — APPROVED** [review]

File: src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java [hunk 16]
```diff
@@ -588,6 +589,10 @@ public Topic getRootCAFilePath() {
                 .addValidator(deTildeValidator);
     }
 
+    public Topic getIpcSocketPath() {
+        return kernel.getConfig().lookup(SYSTEM_NAMESPACE_KEY, DEVICE_PARAM_IPC_SOCKET_PATH);
```
Comment (shaguptashaikh) [inline src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java:593]: I think you want to use `find` instead of `lookup` here

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java [hunk 17]
```diff
@@ -136,6 +144,14 @@ void beforeEach(ExtensionContext context) {
         ignoreExceptionOfType(context, InterruptedException.class);
     }
 
+    @DisabledOnOs(OS.WINDOWS)
+    @Test
+    void Given_assign_path_for_ipcSocket_When_startUp_Then_ipcSocket_store_in_assigned_path() {
+        DeviceConfiguration deviceConfiguration = kernel.getContext().get(DeviceConfiguration.class);
+        String ipcPath = Coerce.toString(deviceConfiguration.getIpcSocketPath());
+        assertThat(new File(ipcPath), is(anExistingFileOrDirectory()));
+        assertEquals(ipcPath, Coerce.toString(kernel.getConfig().getRoot().lookup(SETENV_CONFIG_NAMESPACE, NUCLEUS_DOMAIN_SOCKET_FILEPATH_FOR_COMPONENT)));
```
Comment (shaguptashaikh) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:153]: This tests if you're setting the path correctly for the client (inside the component) to use, what test is testing the server side that creates the socket and sets permissions etc?

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java [hunk 18]
```diff
@@ -136,6 +144,14 @@ void beforeEach(ExtensionContext context) {
         ignoreExceptionOfType(context, InterruptedException.class);
     }
 
+    @DisabledOnOs(OS.WINDOWS)
```
Comment (shaguptashaikh) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:147]: How did we verify this works on Windows?

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java (hunk 17 above)
Comment (MikeDombo) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:153]: The test verifies that the socket exists. The subsequent tests in this class also exercise IPC and they will use the customized path because they are all using the same config file.

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java (hunk 18 above)
Comment (MikeDombo) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:147]: Setting a custom path does _not_ work on Windows. We still use a named pipe as before. The tests run on Windows showing that it is unaffected.

Comment (shaguptashaikh) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:147]: Are there tests that also exercise the default behavior without customized path?

File: src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java (hunk 17 above)
Comment (shaguptashaikh) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:153]: Are there tests that also exercise the default behavior without customized path?

Comment (MikeDombo) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/ipc/IPCServicesTest.java:153]: all of the other tests in the nucleus that use IPC

**Review by shaguptashaikh — COMMENTED** [review]
Please update the config readme with this new config and also explain that it doesn't work for Windows

**Review by shaguptashaikh — APPROVED** [review]

**Comment by MikeDombo** [comment]
@Mirrorxxi  please rebase on main, I would do this for you, but your branch does not allow this.

**Review by MikeDombo — APPROVED** [review]
